### PR TITLE
test(commonTests, alert, combobox, dropdown, input-date-picker, modal, popover, tooltip): extracts a general test utility `OpenClose` for all components implementing `openCloseComponent` util

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #4968

## Summary
Extracts a common test utility `openClose` for all components implementing `openCloseComponent` util. 

Implemented on all relevant components. 

- [x] alert
- [x] combobox
- [x] dropdown
- [x] input-date-picker
- [x] modal
- [x] popover
- [x] tooltip

### CODE REVIEW

File: packages/calcite-components/src/tests/commonTests.ts [hunk 1]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ *
+ * @example
+ * import { emitsOpenCloseTransitionChainedEvents } from "../../tests/commonTests";
+ *
+ * describe("emitsOpenCloseTransitionChainedEvents", () => {
+ *   emitsOpenCloseTransitionChainedEvents("calcite-dropdown", "open", ".container");
+ * });
+ *
+ */
+export function emitsOpenCloseTransitionChainedEvents(
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1340]: Can you simplify the name so it matches the module it supports (similar to other common test helpers)?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 2]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+  componentTestSetup: ComponentTestSetup,
+  toggleProp: string,
+  containerSelector: string
+): void {
+  it(`should emit component status for transition-chained events: 'calciteComponentBeforeOpen', 'calciteComponentOpen', 'calciteComponentBeforeClose', 'calciteComponentClose'`, async () => {
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1345]: A few questions WRT the test name:

* what does `component status` mean?
* what are `transition-chained events`? These aren't mentioned here nor in the module, I believe.
* for readability, can we generalize the event name part to `before and when the component is toggled` (or something along these lines)?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 3]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+    const { page, tag } = await getTagAndPage(componentTestSetup);
+    await page.setContent(`${componentTestSetup}`);
+    const element = await page.find(tag);
+
+    const camelCaseTag = tag.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1350]: I'm not sure what `g` represents, so could you rename `g` to something more expressive? It's the RegExp's match, right?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 4]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+
+    const beforeOpenEvent = page.waitForEvent(`${camelCaseTag}BeforeOpen`);
+    const openEvent = page.waitForEvent(`${camelCaseTag}Open`);
+
+    const beforeOpenSpy = await element.spyOnEvent(`${camelCaseTag}BeforeOpen`);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1355]: Nitpick: can you extract the constructed event names before their use? They're referenced only twice, so I'll leave this up to you.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 5]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ */
+export function emitsOpenCloseTransitionChainedEvents(
+  componentTestSetup: ComponentTestSetup,
+  toggleProp: string,
+  containerSelector: string
+): void {
+  it(`should emit component status for transition-chained events: 'calciteComponentBeforeOpen', 'calciteComponentOpen', 'calciteComponentBeforeClose', 'calciteComponentClose'`, async () => {
+    const { page, tag } = await getTagAndPage(componentTestSetup);
+    await page.setContent(`${componentTestSetup}`);
+    const element = await page.find(tag);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1348]: Disabling animations might improve test robustness.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 6]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+    const beforeCloseEvent = page.waitForEvent(`${camelCaseTag}BeforeClose`);
+    const closeEvent = page.waitForEvent(`${camelCaseTag}Close`);
+
+    const beforeCloseSpy = await element.spyOnEvent(`${camelCaseTag}BeforeClose`);
+    const closeSpy = await element.spyOnEvent(`${camelCaseTag}Close`);
+
+    element.setProperty(toggleProp, true);
+    await page.waitForChanges();
+
+    const container = await page.find(`calcite-dropdown-group >>> ${containerSelector}`);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1367]: Fixme: Avoid hardcoded selectors.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 7]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+    await page.waitForChanges();
+
+    expect(await container.isVisible()).toBe(true);
+
+    await beforeOpenEvent;
+    await openEvent;
+
+    expect(beforeOpenSpy).toHaveReceivedEventTimes(1);
+    expect(openSpy).toHaveReceivedEventTimes(1);
+
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1377]: Can you also assert on the close events not being called?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 8]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+    element.setProperty(toggleProp, false);
+    await page.waitForChanges();
+
+    expect(await container.isVisible()).toBe(false);
+
+    await beforeCloseEvent;
+    await closeEvent;
+
+    expect(beforeCloseSpy).toHaveReceivedEventTimes(1);
+    expect(closeSpy).toHaveReceivedEventTimes(1);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1387]: Can you assert on the open events not being called one more time?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 9]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
     await page.waitForChanges();
   }
 }
+
+/*
+ * @param {ComponentTestSetup} ComponentTestSetup - A component tag, html, or the tag and e2e page for setting up a test.
+ * @param {string} toggleProp - Toggle property to test. Currently, either "open" or "expanded".
+ * @param {string} containerSelector - Container element selector to test for visibility.
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1330]: Testing for visibility seems out of scope for this test helper. `onToggleOpenCloseComponent` helps components easily emit respective open/close events, but doesn't handle visibility, right?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 10]
```diff
@@ -1323,3 +1323,67 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+    const beforeOpenSpy = await element.spyOnEvent(`${camelCaseTag}BeforeOpen`);
+    const openSpy = await element.spyOnEvent(`${camelCaseTag}Open`);
+
+    const beforeCloseEvent = page.waitForEvent(`${camelCaseTag}BeforeClose`);
+    const closeEvent = page.waitForEvent(`${camelCaseTag}Close`);
+
+    const beforeCloseSpy = await element.spyOnEvent(`${camelCaseTag}BeforeClose`);
+    const closeSpy = await element.spyOnEvent(`${camelCaseTag}Close`);
+
+    element.setProperty(toggleProp, true);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1364]: In order to simplify this helper's usage, can you get the initial value of `toggleProp` dynamically and toggle accordingly? Otherwise, this requires the HTML to always be closed. I realize this may be most common setup, but it'd be more flexible to allow both options.

**Review by jcfranco — CHANGES_REQUESTED** [review]
This is looking pretty sweet, @Elijbet! 🍬

Can you wire up this helper in other open/close component test suites?

File: packages/calcite-components/src/tests/commonTests.ts (hunk 5 above)
Comment (jcfranco) [reply packages/calcite-components/src/tests/commonTests.ts:1348]: I forgot to add if you could look into also adding test coverage for when animations are disabled completely? Our animation disabling util technically still allows for transitions to happen (it uses a tiny duration), so you'll need to explicitly set the animation duration to 0.

**Comment by Elijbet** [comment]
> This is looking pretty sweet, @Elijbet! 🍬
> 
> Can you wire up this helper in other open/close component test suites?

Is it ok if I wire up this helper in other components separately from this PR?

I think it's going to bloat up the PR and make it hard to merge. I was hoping to address each case or several cases in a bunch within continuing PRs, so I can progressively adjust the util.

File: packages/calcite-components/src/tests/commonTests.ts (hunk 2 above)
Comment (Elijbet) [reply packages/calcite-components/src/tests/commonTests.ts]: Does this work or did I miss what you meant?

File: packages/calcite-components/src/tests/commonTests.ts (hunk 4 above)
Comment (Elijbet) [reply packages/calcite-components/src/tests/commonTests.ts]: Found a better way!

**Comment by jcfranco** [comment]
> Is it ok if I wire up this helper in other components separately from this PR?

I'd prefer to have the test helper cover all existing use cases from the get go. There's 6 other components that would need to be included, so I don't think it'll add much bloat since we're replacing any existing tests.

File: packages/calcite-components/src/tests/commonTests.ts (hunk 2 above)
Comment (jcfranco) [reply packages/calcite-components/src/tests/commonTests.ts]: Latest test name is perfect, thanks!

File: packages/calcite-components/src/tests/commonTests.ts [hunk 11]
```diff
@@ -1323,3 +1323,63 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ *
+ * @example
+ * import { emitsOpenCloseTransitionChainedEvents } from "../../tests/commonTests";
+ *
+ * describe("emitsOpenCloseTransitionChainedEvents", () => {
+ *   emitsOpenCloseTransitionChainedEvents("calcite-dropdown", "open", ".container");
+ * });
+ *
+ */
+export function emitsOpenCloseEvents(
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1340]: It might be better to avoid implementation specifics in the helper name (namely, `emits` and `events`). This would allow the test helper to cover additional scenarios if needed without having to rename. WDYT about `openClose`? This is going by our current helper naming, 

* `formAssociated`
* `labelable`
* `focusable`
* `reflects`
* `defaults`
* `hidden`
* `accessible`
* `slots`
* `floatingUIOwner`
* `t9n`

File: packages/calcite-components/src/tests/commonTests.ts (hunk 9 above)
Comment (Elijbet) [reply packages/calcite-components/src/tests/commonTests.ts:1330]: Right. We're passing the selector in to `expect(await container.isVisible()).toBe(true); ` before we await for the events. I rewrote it based on some previous comments to pass this instead.     
`emitsOpenCloseEvents(simpleDropdownHTML, "open", `${CSS.calciteDropdownGroup} >>> .${CSS.container}`);`
Does that look ok? I can rewrite that bit to say `Container element selector.`

File: packages/calcite-components/src/tests/commonTests.ts (hunk 11 above)
Comment (Elijbet) [reply packages/calcite-components/src/tests/commonTests.ts:1340]: Sure, what additional scenarios are you thinking about?

Comment (jcfranco) [reply packages/calcite-components/src/tests/commonTests.ts:1340]: I don't have specific scenarios in mind, but if another use case comes up, we don't have to worry about renaming the util.

File: packages/calcite-components/src/components/dropdown/resources.ts [hunk 12]
```diff
@@ -4,4 +4,6 @@ export const SLOTS = {
 
 export const CSS = {
   calciteDropdownContent: "calcite-dropdown-content",
+  calciteDropdownGroup: "calcite-dropdown-group",
```
Comment (jcfranco) [inline packages/calcite-components/src/components/dropdown/resources.ts:7]: These classes don't appear to belong here as they are not used in the component's rendering.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 13]
```diff
@@ -1360,7 +1360,7 @@ export function emitsOpenCloseEvents(
     element.setProperty(toggleProp, true);
     await page.waitForChanges();
 
-    const container = await page.find(`calcite-dropdown-group >>> ${containerSelector}`);
+    const container = await page.find(containerSelector);
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1363]: I think we can focus solely on the events and assume the element will be visible (or not visible) to avoid checking visibility. This should make the helper simpler to use.

File: packages/calcite-components/src/components/modal/modal.e2e.ts [hunk 14]
```diff
@@ -13,6 +13,14 @@ describe("calcite-modal properties", () => {
     hidden("calcite-modal");
   });
 
+  describe("openClose", () => {
+    openClose("<calcite-modal></calcite-modal>", "open", false);
```
Comment (jcfranco) [inline packages/calcite-components/src/components/modal/modal.e2e.ts:17]: If you update `openClose` to use `simplePageSetup` (similar to other test helpers), it will allow devs to set up a test by passing the tag directly in this case. 

```ts
openClose("calcite-modal", "open", false);
```

File: packages/calcite-components/src/tests/commonTests.ts [hunk 15]
```diff
@@ -1341,3 +1345,124 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ *  openTooltip: (page: E2EPage) => Promise<void>;
+ *  closeTooltip: (page: E2EPage) => Promise<void>;
+ * }} userInputDevice - Optional argument with functions to simulate user input (mouse or keyboard), to open or close the component.
+ */
+
+export function openClose(
+  componentTestSetup: ComponentTestSetup,
+  toggleProp: string,
+  toggleValue: boolean,
+  userInputDevice?: {
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1381]: Is there another name we could use for this? Not sure I would understand what `userInputDevice` would be used for without going into the code.

Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1381]: Can you look at introducing an `options` object? If you look at other test utils, additional props not required for the test usually go in an `options` argument.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 16]
```diff
@@ -1341,3 +1345,124 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ * @param {ComponentTestSetup} ComponentTestSetup - A component tag, html, or the tag and e2e page for setting up a test.
+ * @param {string} toggleProp - Toggle property to test. Currently, either "open" or "expanded".
+ * @param {boolean} toggleValue - Indicates the initial value of the toggle property.
+ * @param {{
+ *  openTooltip: (page: E2EPage) => Promise<void>;
+ *  closeTooltip: (page: E2EPage) => Promise<void>;
+ * }} userInputDevice - Optional argument with functions to simulate user input (mouse or keyboard), to open or close the component.
+ */
+
+export function openClose(
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1377]: Can you make sure the the body of this function is typed as much as possible?

File: packages/calcite-components/src/tests/commonTests.ts [hunk 17]
```diff
@@ -30,9 +30,13 @@ function isHTML(tagOrHTML: string): boolean {
 
 function getTag(tagOrHTML: string): ComponentTag {
   if (isHTML(tagOrHTML)) {
-    const regex = /[>\s]/;
+    const regex = /<calcite-[a-z0-9-]+/i;
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:33]: * `regex` was there already, but it might be good to rename it to express what it'll match.
* `match` could also be renamed to better convey its purpose.
* The error message could have more info on why it didn't match (e.g., no calcite-component tag used). Also, WDYT about throwing if it finds more than 1? Might not happen much, but if it does, it'll be useful to have this early warning when using the helper.

**Review by jcfranco — COMMENTED** [review]
Looking awesome, @Elijbet! 😎

**Comment by Elijbet** [comment]
Cool, so far mostly syntax comments. I'll work on that and modal failing tests next.

**Comment by Elijbet** [comment]
This should resolve the failing modal test issue #7273.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 18]
```diff
@@ -1376,3 +1380,165 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ */
+
+type userInputDevice = {
+  open: (page: E2EPage) => Promise<void>;
+  close: (page: E2EPage) => Promise<void>;
+};
+
+export async function openClose(
+  componentTagOrHTML: TagOrHTML,
+  toggleProp = "open",
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1416]: `open` is the default, but it is still being specified in all tests. 

Most of our test helpers provide arguments for required values, and extra options are provided in an options parameters. I think we should do this to simplify this helper's usage:

```ts
interface OpenCloseOptions {
  /**
   * ...
   */
  initialToggleValue?: boolean;
  
  /**
   * ...
   */
  toggleProp?: string;

  /**
   * ...
   */
  userInputDevice?: UserInputDevice;
}

export async function openClose(
  componentTagOrHTML: TagOrHTML,
  options?: OpenCloseOptions
)
```

The main benefit we get from this is that we avoid passing defaults when we need to use `userInputDevice`:

```ts
openClose("calcite-magic", {
  userInputDevice: { /* ... */ }
});
```

vs

```ts
openClose("calcite-magic", "open", false, {
  userInputDevice: { /* ... */ }
});
```

File: packages/calcite-components/src/tests/commonTests.ts [hunk 19]
```diff
@@ -1376,3 +1380,165 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+ *   },
+ * });
+ *
+ * @param {string} componentTagOrHTML - the component tag or HTML markup to test against
+ * @param {string} toggleProp - Toggle property to test. Currently, either "open" or "expanded".
+ * @param {boolean} initialToggleValue - Indicates the initial value of the toggle property to determine whether to configure a `simplePageSetup` or `newProgrammaticE2EPage`.
+ * @param {userInputDevice} userInputDevice - Optional argument with functions to simulate user input (mouse or keyboard), to open or close the component.
+ */
+
+type userInputDevice = {
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1409]: Please add JSDoc to the interface similar to other test util options. Applies to all test-helper-related types/interfaces.

File: packages/calcite-components/src/tests/commonTests.ts (hunk 18 above)
Comment (jcfranco) [reply packages/calcite-components/src/tests/commonTests.ts:1416]: On a related note, I think some parameter names could be renamed to better convey their intent. This is just me, but this is how I'm currently reading them:

* `toggleProp` – most likely from `toggle`, but this can read as a boolean to enable toggling
* `initialToggleValue` – the initial value before toggling takes place or to toggle with
* `userInputDevice` – an object to specify how the user will interact with the component

Let's sync up on this to hash out some names.

File: packages/calcite-components/src/components/tooltip/tooltip.e2e.ts [hunk 20]
```diff
@@ -680,7 +762,7 @@ describe("calcite-tooltip", () => {
   describe("beforeOpen, open, beforeClose, close event emitting", () => {
```
Comment (jcfranco) [inline packages/calcite-components/src/components/tooltip/tooltip.e2e.ts:762]: Can these be replaced by using the helper?

File: packages/calcite-components/src/components/action-bar/action-bar.tsx [hunk 21]
```diff
@@ -407,8 +407,8 @@ export class ActionBar
     return (
       <calcite-action-group
         class={CSS.actionGroupEnd}
-        label={actionsEndGroupLabel}
         hidden={this.expandDisabled && !(this.hasActionsEnd || this.hasBottomActions)}
+        label={actionsEndGroupLabel}
```
[the remark this answers is not in the document]
Comment (Elijbet) [reply packages/calcite-components/src/components/action-bar/action-bar.tsx:411]: Not sure how this happened TBH, and now I think have to add the `eslint-disable` for it to move back.

File: packages/calcite-components/src/tests/commonTests.ts [hunk 22]
```diff
@@ -1376,3 +1380,204 @@ export async function t9n(componentTestSetup: ComponentTestSetup): Promise<void>
...
+interface OpenCloseOptions {
+  /**
+   * Toggle property to test. Currently, either "open" or "expanded".
+   */
+  openPropName?: string;
+
+  /**
+   * Indicates the initial value of the toggle property to determine whether to configure a `simplePageSetup` or `newProgrammaticE2EPage`.
+   */
+  initialToggleValue?: boolean;
```
Comment (jcfranco) [inline packages/calcite-components/src/tests/commonTests.ts:1405]: Synced up w/ @Elijbet and we decided to see if we could eliminate the need for this prop by looking up the open prop name dynamically from the test HTML or component (separate PR).

**Review by jcfranco — APPROVED** [review]
🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧
🔧🔧🧪🧪🧪🔧🔧🧪🧪🔧🔧🔧🧪🧪🔧🔧🧪🔧🔧🔧🧪🔧
🔧🧪🔧🔧🔧🔧🧪🔧🔧🧪🔧🧪🔧🔧🧪🔧🧪🔧🔧🔧🧪🔧
🔧🧪🔧🔧🔧🔧🧪🔧🔧🧪🔧🧪🔧🔧🧪🔧🧪🔧🔧🔧🧪🔧
🔧🧪🔧🔧🔧🔧🧪🔧🔧🧪🔧🧪🔧🔧🧪🔧🧪🔧🔧🔧🔧🔧
🔧🔧🧪🧪🧪🔧🔧🧪🧪🔧🔧🔧🧪🧪🔧🔧🧪🧪🧪🔧🧪🔧
🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧🔧
